File: solutions/C++/add_binary_string_simulation.cpp
#include <string>
#include <iostream>

using namespace std;

string addBinaryStringSimulation(string a, string b) {
    /*
    String simulation approach - simulate binary addition from right to left.
    Time: O(max(a.size(), b.size()))
    Space: O(max(a.size(), b.size())) for result
    */
    string result = "";
    int carry = 0;
    int i = a.size() - 1;
    int j = b.size() - 1;

    while (i >= 0 || j >= 0 || carry) {
        int digit_a = (i >= 0) ? (a[i] - '0') : 0;
        int digit_b = (j >= 0) ? (b[j] - '0') : 0;

        int total = digit_a + digit_b + carry;
        result = char('0' + (total % 2)) + result;
        carry = total / 2;

        i--;
        j--;
    }

    return result;
}

int main() {
    cout << addBinaryStringSimulation("11", "1") << endl;      // "100"
    cout << addBinaryStringSimulation("1010", "1011") << endl; // "10101"
    cout << addBinaryStringSimulation("0", "0") << endl;       // "0"
    return 0;
}

File: solutions/Rust/add_binary_string_simulation.rs
pub fn add_binary_string_simulation(a: String, b: String) -> String {
    /*
    String simulation approach - simulate binary addition from right to left.
    Time: O(max(a.len(), b.len()))
    Space: O(max(a.len(), b.len())) for result
    */
    let mut result = Vec::new();
    let mut carry = 0;
    let a_bytes = a.as_bytes();
    let b_bytes = b.as_bytes();
    let mut i = a_bytes.len() as i32 - 1;
    let mut j = b_bytes.len() as i32 - 1;

    while i >= 0 || j >= 0 || carry > 0 {
        let digit_a = if i >= 0 { (a_bytes[i as usize] - b'0') as i32 } else { 0 };
        let digit_b = if j >= 0 { (b_bytes[j as usize] - b'0') as i32 } else { 0 };

        let total = digit_a + digit_b + carry;
        result.push((b'0' + (total % 2) as u8) as char);
        carry = total / 2;

        i -= 1;
        j -= 1;
    }

    result.reverse();
    result.iter().collect()
}

fn main() {
    println!("{}", add_binary_string_simulation("11".to_string(), "1".to_string()));      // "100"
    println!("{}", add_binary_string_simulation("1010".to_string(), "1011".to_string())); // "10101"
    println!("{}", add_binary_string_simulation("0".to_string(), "0".to_string()));       // "0"
}
